File: src/browser-context.mjs
// Display-only envelope removal. Never rewrite the agent's canonical prompt.
export function stripSidebarBrowserContext(value) {
  const text = String(value ?? '');
  return text.replace(/(?:^|\r?\n\r?\n)<browser-context>([\s\S]*?)<\/browser-context>(?=\s*(?:$|Attached files \())/g, (block, json, offset) => {
    // Preserve fenced examples and malformed/unrelated user content.
    let fence = null;
    for (const line of text.slice(0, offset).split(/\r?\n/)) {
      const m = /^ {0,3}(`{3,}|~{3,})(.*)$/.exec(line);
      if (m && !fence) fence = m[1];
      else if (m && m[1][0] === fence?.[0] && m[1].length >= fence.length && !m[2].trim()) fence = null;
    }
    if (fence) return block;
    try {
      const c = JSON.parse(json);
      const valid = id => typeof id === 'string' && /^control-target:chrome-tab:[a-zA-Z0-9-]+$/.test(id);
      return c.source === 'Zyra Chrome sidebar' && (valid(c.targetId) || (Array.isArray(c.targets) && c.targets.length && c.targets.every(t => valid(t?.targetId)))) ? '' : block;
    } catch { return block; }
  });
}

File: mobile/android/app/src/main/java/dev/zyra/mobile/data/BrowserContext.kt
package dev.zyra.mobile.data

import org.json.JSONObject

/** Display-only; leave canonical/raw envelopes untouched for the agent. */
object BrowserContext {
    private val block = Regex("(?:^|\\r?\\n\\r?\\n)<browser-context>([\\s\\S]*?)</browser-context>(?=\\s*(?:$|Attached files \\())")
    fun display(text: String): String = block.replace(text) { match ->
        var fence: String? = null
        text.take(match.range.first).lineSequence().forEach { line ->
            Regex("^ {0,3}(`{3,}|~{3,})(.*)$").matchEntire(line)?.let { f ->
                val run = f.groupValues[1]; val previous = fence
                if (previous == null) fence = run
                else if (run.first() == previous.first() && run.length >= previous.length && f.groupValues[2].isBlank()) fence = null
            }
        }
        val valid = runCatching {
            val c = JSONObject(match.groupValues[1])
            fun target(id: String) = Regex("control-target:chrome-tab:[a-zA-Z0-9-]+").matches(id)
            val targets = c.optJSONArray("targets")
            c.optString("source") == "Zyra Chrome sidebar" && (target(c.optString("targetId")) || targets != null && targets.length() > 0 && (0 until targets.length()).all { target(targets.optJSONObject(it)?.optString("targetId").orEmpty()) })
        }.getOrDefault(false)
        if (fence == null && valid) "" else match.value
    }
}
